File: WebClient SDK/src/default/ui/krGetCycleFileInfoUIExtension.ts
import { CardUIExtension, ICardUIExtensionContext } from 'tessa/ui/cards';
import { tryGetFromInfo, createCardFileSourceForCard } from 'tessa/ui';
import { IStorage, clear } from 'tessa/platform/storage';
import { Guid, TypedField, DotNetType, getTypedFieldValue, createTypedField } from 'tessa/platform';
import { VirtualFile, VirtualFileVersion } from 'tessa/files';
import { CardFileContentSource } from 'tessa/cards';

export class KrGetCycleFileInfoUIExtension extends CardUIExtension {
  public async initializing(context: ICardUIExtensionContext) {
    if (context.model.inSpecialMode) {
      return;
    }

    const cardInfo = context.card.tryGetInfo();
    if (!cardInfo) {
      return;
    }

    const filesByCycles = tryGetFromInfo<{ [key: string]: TypedField<DotNetType.Guid, guid> }>(
      cardInfo,
      'KrFilesByCycles'
    );
    const filesModifiedByCyclesStorage = tryGetFromInfo<IStorage[]>(
      cardInfo,
      'KrFilesModifiedByCycles'
    );
    const maxCycleNumber = tryGetFromInfo<number>(cardInfo, 'KrMaxCycleNumber');
    // tslint:disable-next-line: triple-equals
    if (!filesByCycles || !filesModifiedByCyclesStorage || maxCycleNumber == undefined) {
      return;
    }

    for (let fileId in filesByCycles) {
      const cycleId = getTypedFieldValue(filesByCycles[fileId]);
      const fileModel = context.fileContainer.files.find(x => Guid.equals(x.id, fileId));
      if (fileModel) {
        fileModel.info['KrCycleID'] = Number(cycleId);
        fileModel.info['KrCycleorder'] = Number(maxCycleNumber) - Number(cycleId);
        fileModel.info['KrMaxCycleNumber'] = Number(maxCycleNumber);
      }
    }

    if (filesModifiedByCyclesStorage.length > 0) {
      // clone - пустая карточка с той же информацией по типу и по версии, и др. системной информацией, но без фактических данных;
      // в неё будут добавляться виртуальные файлы, чтобы незахламлять структуру основной карточки

      const clone = context.card.clone();
      clone.sections.clear();
      clone.files.clear();
      clone.tasks.clear();
      clone.taskHistory.clear();
      clone.taskHistoryGroups.clear();
      clear(clone.info);

      const cloneFileSource = createCardFileSourceForCard(clone);

      for (let versionInfo of filesModifiedByCyclesStorage) {
        const fileId = tryGetFromInfo<guid>(versionInfo, 'FileID');

        const originalFile = context.fileContainer.files.find(x => Guid.equals(x.id, fileId));
        if (!originalFile) {
          continue;
        }

        const originalCardFile = context.card.files.find(x => Guid.equals(x.rowId, fileId));
        if (!originalCardFile) {
          continue;
        }

        const cycle = tryGetFromInfo<number>(versionInfo, 'Cycle');
        const versionId = tryGetFromInfo<guid>(versionInfo, 'VersionID');
        const versionNumber = tryGetFromInfo<number>(versionInfo, 'Number');
        const versionSize = tryGetFromInfo<number>(versionInfo, 'Size');
        const sourceId = tryGetFromInfo<number>(versionInfo, 'SourceID');
        const versionCreated = tryGetFromInfo<string>(versionInfo, 'Created');
        const versionCreatedById = tryGetFromInfo<guid>(versionInfo, 'CreatedByID');
        const versionCreatedByName = tryGetFromInfo<string>(versionInfo, 'CreatedByName');

        // теперь создаём файл и наполняем его
        const virtualFile = await context.fileContainer.addVirtalFile(
          cloneFileSource,
          new VirtualFile(
            originalCardFile.typeName,
            Guid.newGuid(),
            originalCardFile.name,
            token => {
              token.size = versionSize;
              token.lastVersionTags.push(...originalFile.lastVersion.tags);
            }
          ),
          new VirtualFileVersion(Guid.newGuid(), originalCardFile.name, token => {
            token.size = versionSize;
            token.number = versionNumber;
            token.created = versionCreated;
            token.createdById = versionCreatedById;
            token.createdByName = versionCreatedByName;
            token.tags.push(...originalFile.lastVersion.tags);
          })
        );

        const virtualCardFile = clone.files.find(x => x.card.id === virtualFile.id)!;
        virtualCardFile.card.createdById = versionCreatedById;
        virtualCardFile.card.createdByName = versionCreatedByName;
        virtualCardFile.card.created = versionCreated;
        virtualCardFile.lastVersion!.number = versionNumber;
        virtualCardFile.lastVersion!.created = versionCreated;
        virtualCardFile.lastVersion!.createdById = versionCreatedById;
        virtualCardFile.lastVersion!.createdByName = versionCreatedByName;

        virtualCardFile.externalSource = new CardFileContentSource();
        virtualCardFile.externalSource.cardId = context.card.id;
        virtualCardFile.externalSource.cardTypeId = context.model.cardType.id!;
        virtualCardFile.externalSource.fileId = fileId;
        virtualCardFile.externalSource.storeSource = sourceId;
        virtualCardFile.externalSource.versionRowId = versionId;

        virtualFile.info['KrCreated'] = createTypedField(versionCreated, DotNetType.DateTime);
        virtualFile.info['KrCreatedByName'] = createTypedField(
          versionCreatedByName,
          DotNetType.String
        );
        virtualFile.info['KrCycleID'] = createTypedField(cycle, DotNetType.Int32);
        virtualFile.info['KrCycleorder'] = createTypedField(
          maxCycleNumber - cycle,
          DotNetType.Int32
        );
        virtualFile.info['KrMaxCycleNumber'] = createTypedField(maxCycleNumber, DotNetType.Int32);
      }
    }
  }
}

File: Source/Extensions/Tessa.Extensions.Default.Client/UI/KrGetCycleFileInfoUIExtension.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Tessa.Cards;
using Tessa.Extensions.Default.Shared;
using Tessa.Files;
using Tessa.Platform.Storage;
using Tessa.UI.Cards;

namespace Tessa.Extensions.Default.Client.UI
{
    public sealed class KrGetCycleFileInfoUIExtension : CardUIExtension
    {
        #region Constructors

        public KrGetCycleFileInfoUIExtension(CreateFileSourceForCardFuncAsync createFileSourceForCardFuncAsync) =>
            this.createFileSourceForCardFuncAsync = createFileSourceForCardFuncAsync ?? throw new ArgumentNullException(nameof(createFileSourceForCardFuncAsync));

        #endregion

        #region Fields

        private readonly CreateFileSourceForCardFuncAsync createFileSourceForCardFuncAsync;

        #endregion

        #region Base Overrides

        public override async Task Initializing(ICardUIExtensionContext context)
        {
            if (context.Model.InSpecialMode())
            {
                return;
            }

            Dictionary<string, object> cardInfo = context.Card.TryGetInfo();

            if (cardInfo is null
                || !cardInfo.TryGetValue(CycleGroupingInfoKeys.FilesByCyclesKey, out object filesByCyclesObj)
                || !(filesByCyclesObj is Dictionary<string, object> filesByCycles)
                || !cardInfo.TryGetValue(CycleGroupingInfoKeys.FilesModifiedByCyclesKey, out object filesModifiedByCyclesObj)
                || !(filesModifiedByCyclesObj is IList filesModifiedByCyclesStorage)
                || !cardInfo.TryGetValue(CycleGroupingInfoKeys.MaxCycleNumberKey, out object maxCycleNumberObj))
            {
                return;
            }

            foreach ((string fileID, object cycleID) in filesByCycles)
            {
                IFile fileModel = context.FileContainer.Files.TryGet(Guid.Parse(fileID));
                if (fileModel != null)
                {
                    fileModel.Info[CycleGroupingInfoKeys.CycleIDKey] = (int) cycleID;
                    fileModel.Info[CycleGroupingInfoKeys.CycleOrderKey] = (int) maxCycleNumberObj - (int) cycleID;
                    fileModel.Info[CycleGroupingInfoKeys.MaxCycleNumberKey] = (int) maxCycleNumberObj;
                }
            }

            if (filesModifiedByCyclesStorage.Count > 0)
            {
                // clone - пустая карточка с той же информацией по типу и по версии, и др. системной информацией, но без фактических данных;
                // в неё будут добавляться виртуальные файлы, чтобы незахламлять структуру основной карточки

                Card clone = context.Card.Clone();
                clone.Sections.Clear();
                clone.Files.Clear();
                clone.Tasks.Clear();
                clone.TaskHistory.Clear();
                clone.TaskHistoryGroups.Clear();
                clone.Info.Clear();

                FileSourceForCard cloneFileSource = await this.createFileSourceForCardFuncAsync(clone, context.CancellationToken);

                foreach (Dictionary<string, object> versionInfo in filesModifiedByCyclesStorage)
                {
                    Guid fileID = versionInfo.Get<Guid>("FileID");

                    IFile originalFile = context.FileContainer.Files.FirstOrDefault(p => p.ID == fileID);
                    if (originalFile is null)
                    {
                        continue;
                    }

                    CardFile originalCardFile = context.Card.Files.FirstOrDefault(p => p.RowID == fileID);
                    if (originalCardFile is null)
                    {
                        continue;
                    }

                    // сначала читаем и кастим все данные
                    int cycle = versionInfo.Get<int>("Cycle");
                    Guid versionID = versionInfo.Get<Guid>("VersionID");
                    int versionNumber = versionInfo.Get<int>("Number");
                    long versionSize = versionInfo.Get<long>("Size");
                    int sourceID = versionInfo.Get<int>("SourceID");
                    DateTime versionCreated = versionInfo.Get<DateTime>("Created");
                    Guid versionCreatedByID = versionInfo.Get<Guid>("CreatedByID");
                    string versionCreatedByName = versionInfo.Get<string>("CreatedByName");

                    // теперь создаём файл и наполняем его
                    IFile virtualFile =
                        await context.FileContainer.AddVirtualAsync(
                            cloneFileSource,
                            new VirtualFile(originalCardFile.TypeName, originalCardFile.Name,
                                async (token, ct) =>
                                {
                                    token.Size = versionSize;
                                    token.LastVersionTags.AddRange(originalFile.Versions.Last.Tags);
                                }),
                            cancellationToken: context.CancellationToken,
                            versions: new []
                            {
                                new VirtualFileVersion(originalCardFile.Name,
                                    async (token, ct) =>
                                    {
                                        token.Size = versionSize;
                                        token.Number = versionNumber;
                                        token.Created = versionCreated;
                                        token.CreatedByID = versionCreatedByID;
                                        token.CreatedByName = versionCreatedByName;
                                        token.Tags.AddRange(originalFile.Versions.Last.Tags);
                                    })    
                            });

                    CardFile virtualCardFile = clone.Files.First(p => p.Card.ID == virtualFile.ID);
                    virtualCardFile.Card.CreatedByID = versionCreatedByID;
                    virtualCardFile.Card.CreatedByName = versionCreatedByName;
                    virtualCardFile.Card.Created = versionCreated;
                    virtualCardFile.LastVersion.Number = versionNumber;
                    virtualCardFile.LastVersion.Created = versionCreated;
                    virtualCardFile.LastVersion.CreatedByID = versionCreatedByID;
                    virtualCardFile.LastVersion.CreatedByName = versionCreatedByName;

                    virtualCardFile.ExternalSource = new CardFileContentSource
                    {
                        CardID = context.Card.ID,
                        CardTypeID = context.Model.CardType.ID,
                        FileID = fileID,
                        Source = new CardFileSourceType(sourceID),
                        VersionRowID = versionID,
                    };

                    virtualFile.Info[CycleGroupingInfoKeys.CreatedKey] = versionCreated;
                    virtualFile.Info[CycleGroupingInfoKeys.CreatedByNameKey] = versionCreatedByName;
                    virtualFile.Info[CycleGroupingInfoKeys.CycleIDKey] = cycle;
                    virtualFile.Info[CycleGroupingInfoKeys.CycleOrderKey] = (int) maxCycleNumberObj - cycle;
                    virtualFile.Info[CycleGroupingInfoKeys.MaxCycleNumberKey] = (int)maxCycleNumberObj;
                }
            }
        }

        #endregion
    }
}